Add sonar beamformers

PhilClemson opened the pull request:
This PR adds two new detectors for application to sonar tracking problems. Both take hydrophone data and convert these signals to a set of detections using a windowing method.

`capon` is a standard Capon beamformer. `rjmcmc` is a novel beamforming algorithm based on reversible-jump Markov chain Monte Carlo. The latter algorithm is designed for multiple target tracking, but the Metropolis-Hastings algorithm currently used does not perform well in these situations. I've therefore disabled the multiple target detection for now by forcing the update type in `proposal_func()`. The current algorithm has since been replaced with a more effective no-u-turn sampler (NUTS) but this is still undergoing testing (I'll include this in another PR at a later date).

The usage is demonstrated in the corresponding test file `test_beamformers.py`, where I've included a small simulated dataset `fixed_target_example.csv`.

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,483 @@
+import numpy as np
+import math
+import random
+import numba
```
I'm not sure we want to include `numba` dependency for the one function below. Suggest removing this for now.

Separately I'd be interested if we could use `numba` in certain places however.

PhilClemson replied:
Good spot @sdhiscocks, I'd mistakenly left that function in from an earlier version which was doing some complex number operations. I've removed both `numba` and the function now.

sdhiscocks commented on stonesoup/detector/tests/test_beamformers.py:
```diff
@@ -0,0 +1,64 @@
+from stonesoup.types.track import Track
+
+
+def fixed_target_beamformer_test():
```
Looks like this didn't run. Test functions must start with `test_`
```suggestion
def test_fixed_target_beamformer():
```

PhilClemson replied:
I've now corrected this (sorry, not familiar with PyTest).

I also noticed that I'd left in the hardcoded value of the signal frequency for testing, which I've now removed.

PhilClemson commented:
I have now added some descriptions / references above the beamformer classes. I also added a user-set random seed so that the RJMCMC algorithm will give reproducible output (for testing).

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,505 @@
+
+        N = self.num_sensors*window
+
+        max_targets = 5
```
Should this be configurable?

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,505 @@
+        phivals = np.linspace(0, math.pi/2, num=100)
+
+        # spatial locations of hydrophones
+        z = np.matrix(self.sensor_loc)
```
`np.matrix` is deprecated. Should use `np.array` instead.

sdhiscocks commented on stonesoup/detector/tests/test_beamformers.py:
```diff
@@ -0,0 +1,64 @@
+from stonesoup.types.track import Track
+
+
+def test_fixed_target_beamformer():
```
Could parameterise this test:

```patch
diff --git a/stonesoup/detector/tests/test_beamformers.py b/stonesoup/detector/tests/test_beamformers.py
index fb990f2b..312d31ab 100644
--- a/stonesoup/detector/tests/test_beamformers.py
+++ b/stonesoup/detector/tests/test_beamformers.py
@@ -1,8 +1,10 @@
 # -*- coding: utf-8 -*-
-import numpy as np
 from os import path
 from datetime import datetime
 
+import numpy as np
+import pytest
+
 from stonesoup.detector import beamformers
 from stonesoup.models.transition.linear import CombinedLinearGaussianTransitionModel, \
                                                ConstantVelocity
@@ -15,9 +17,16 @@ from stonesoup.types.hypothesis import SingleHypothesis
 from stonesoup.types.track import Track
 
 
-def test_fixed_target_beamformer():
-
+@pytest.fixture(params=[beamformers.capon, beamformers.rjmcmc])
+def detector(request):
     data_file = path.join(path.dirname(__file__), "fixed_target_example.csv")
+    extra_args = {'seed': 12345} if request.param is beamformers.rjmcmc else {}
+    return request.param(data_file, sensor_loc="0 0 0; 0 10 0; 0 20 0; 10 0 0; 10 10 0; \
+        10 20 0; 20 0 0; 20 10 0; 20 20 0", fs=2000, omega=50, wave_speed=1481, **extra_args)
+
+
+def test_fixed_target_beamformer(detector):
+
     truth = [0.8, 0.2]
 
     transition_model = CombinedLinearGaussianTransitionModel([ConstantVelocity(0.01),
@@ -32,27 +41,9 @@ def test_fixed_target_beamformer():
 
     prior = GaussianState([[0.5], [0], [0.5], [0]], np.diag([1, 0, 1, 0]),
                           timestamp=datetime.now())
+    track = Track()
 
-    detector1 = beamformers.capon(data_file, sensor_loc="0 0 0; 0 10 0; 0 20 0; 10 0 0; 10 10 0; \
-        10 20 0; 20 0 0; 20 10 0; 20 20 0", fs=2000, omega=50, wave_speed=1481)
-    detector2 = beamformers.rjmcmc(data_file, sensor_loc="0 0 0; 0 10 0; 0 20 0; 10 0 0; 10 10 0; \
-        10 20 0; 20 0 0; 20 10 0; 20 20 0", fs=2000, omega=50, wave_speed=1481, seed=12345)
-
-    track1 = Track()
-    track2 = Track()
-
-    for timestep, detections in detector2:
-        for detection in detections:
-            # check beamformer accuracy
-            assert(np.abs(detection.state_vector[0] - truth[0]) < 0.4)
-            assert(np.abs(detection.state_vector[1] - truth[1]) < 0.1)
-            prediction = predictor.predict(prior, timestamp=detection.timestamp)
-            hypothesis = SingleHypothesis(prediction, detection)  # Group prediction + measurement
-            post = updater.update(hypothesis)
-            track2.append(post)
-            prior = track2[-1]
-
-    for timestep, detections in detector1:
+    for timestep, detections in detector:
         for detection in detections:
             # check beamformer accuracy
             assert(np.abs(detection.state_vector[0] - truth[0]) < 0.4)
@@ -60,5 +51,5 @@ def test_fixed_target_beamformer():
             prediction = predictor.predict(prior, timestamp=detection.timestamp)
             hypothesis = SingleHypothesis(prediction, detection)  # Group prediction + measurement
             post = updater.update(hypothesis)
-            track1.append(post)
-            prior = track1[-1]
+            track.append(post)
+            prior = track[-1]
```

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,505 @@
+    csv_path: str = Property(doc='The path to the csv file, containing the raw data')
+    fs: float = Property(doc='Sampling frequency (Hz)')
+    omega: float = Property(doc='Signal frequency (Hz)')
+    sensor_loc: str = Property(doc='Cartesian coordinates of the sensors in the format\
```
I think a `Sequence[StateVector]` would be better for both beamformers.

PhilClemson replied:
Just trying to understand this - is the issue that I'm passing the csv file / data directly into the detector rather than using a reader and passing everything as a `Sequence[StateVector]`?

sdhiscocks replied:
It's the fact that your passing in sensor locations as a string of numbers, rather than just passing the numbers as an array, or list of arrays for example.

PhilClemson replied:
Got it, thanks! This has been updated in the latest version.

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,505 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+import math
+import random
+import copy
+from datetime import datetime, timedelta
+from stonesoup.base import Property, Base
+from stonesoup.buffered_generator import BufferedGenerator
+from stonesoup.models.measurement.linear import LinearGaussian
+from stonesoup.types.array import StateVector, CovarianceMatrix
+from stonesoup.types.detection import Detection
```
```suggestion
import copy
import math
import random
from datetime import datetime, timedelta

import numpy as np

from ..base import Property, Base
from ..buffered_generator import BufferedGenerator
from ..models.measurement.linear import LinearGaussian
from ..types.array import StateVector, CovarianceMatrix
from ..types.detection import Detection
```

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,505 @@
+from stonesoup.types.detection import Detection
+
+
+def proposal(params, K, p_params):
```
Are these standalone functions only used by `rjmcmc`? If so, I'd add them as class/static methods to the `rjmcmc` class.

sdhiscocks reviewed (changes requested):
I'm not familiar with beamforming, but overall looks good.

Code coverage could be improved with some additional tests if possible. Parameterising the tests may make this easier. But may be limited by test data you have available?

This is required to build documentation:
```patch
diff --git a/docs/source/stonesoup.detector.rst b/docs/source/stonesoup.detector.rst
index f0e86d4b..5a3f3e74 100644
--- a/docs/source/stonesoup.detector.rst
+++ b/docs/source/stonesoup.detector.rst
@@ -10,4 +10,8 @@ Detectors
 .. automodule:: stonesoup.detector.tensorflow
     :show-inheritance:
 
+Beamformers
+-----------
 
+.. automodule:: stonesoup.detector.beamformers
+    :show-inheritance:

```

oharrald-Dstl commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,544 @@
+    path: str = Property(doc='The path to the csv file, containing the raw data')
+    fs: float = Property(doc='Sampling frequency (Hz)')
+    omega: float = Property(doc='Signal frequency (Hz)')
+    sensor_loc: list = Property(doc='Cartesian coordinates of the sensors in the format\
```
For `sensor_loc`, having a list of state vectors might be better `Sequence[StateVector]` (as mentioned by SH... I think)

oharrald-Dstl commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,544 @@
+
+                yield current_time, {detection}
+
+    def log_prob(self, p_noise, p_params, p_K, y, T, sinTy, cosTy, yTy, sumsinsq, sumcossq,
```
Some documentation of this method would be good (and others in this class).

oharrald-Dstl commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,544 @@
+                p_params[k, 1] = p_params[k, 1] + math.pi
+
+            # wrap angles
+            if p_params[k, 0] < 0:
+                p_params[k, 0] = p_params[k, 0] + 2*math.pi
+            elif p_params[k, 0] > 2*math.pi:
+                p_params[k, 0] = p_params[k, 0] - 2*math.pi
+            if p_params[k, 1] < 0:
+                p_params[k, 1] = p_params[k, 1] + 2*math.pi
+            elif p_params[k, 1] > 2*math.pi:
+                p_params[k, 1] = p_params[k, 1] - 2*math.pi
```
There are `Bearing` and `Elevation` types that could be used for this.

PhilClemson replied:
Thanks for this, I wasn't aware these existed. I've updated the parameter arrays to be StateVector[Elevation, Bearing]. However, I don't think I can use the `mod_angle` functions (at least for the elevation) as it's tied in with the MCMC algorithm, which works on the range [0, 2pi] for all angles due to the requirements of the method.

oharrald-Dstl commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,544 @@
+    def noise_proposal(self, noise):
+        epsilon = random.gauss(0, 0.1)
+        rand_val = abs(noise+epsilon)
+        p_noise = copy.deepcopy(rand_val)
```
Why is a deep copy necessary here?

oharrald-Dstl commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,544 @@
+            path = Path(path)
+        super().__init__(path, *args, **kwargs)
+        if self.seed != 0:
+            random.seed(a=self.seed, version=2)
```
This will set the global random seed, fixing output for entire simulation.
Consider method used in #444 and #461 for locally fixing random seed.

PhilClemson commented:
Okay, I think I've made all the requested changes but let me know if I missed anything!

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,560 @@
+                    [unique_peak_inds1, unique_peak_inds2] = np.unravel_index(
+                                                             param_hist[0, :, :].argmax(),
+                                                             param_hist[0, :, :].shape)
+                    num_peaks = 1
```
Variable isn't used.

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,560 @@
+
+                yield current_time, {detection}
+
+    def log_prob(self, p_noise, p_params, p_K, y, T, sinTy, cosTy, yTy, sumsinsq, sumcossq,
```
`p_noise`, `y` and `T` aren't used

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,560 @@
+        removing a target).
+
+        """
+        update_type = uniform.rvs(random_state=self.random_state)
+        p_K = 0
+        Qratio = 1  # ratio of proposal probabilities for forwards and backwards moves
+        update_type = 1  # forced temporarily (for single-target examples)
```
`update_type` is defined twice.

sdhiscocks commented on stonesoup/detector/beamformers.py:
```diff
@@ -0,0 +1,560 @@
+        proposals centred on the current parameter values.
+
+        """
+        p_K = 0
```
Not required

PhilClemson commented:
Thanks again for looking at this @sdhiscocks and apologies for taking so long to get back to you with the changes (unfortunately Christmas got in the way and then I was busy with other work in January).

I think I've fixed all the issues that were mentioned but it might be worth checking the new reader implementation with `itertools` is usable for streaming data.
